README.rst
===================
edw.seleniumtesting
===================

Selenium based automated testing.


Installation
------------
This package requires **Python 3.5**!
::

    $ pip install edw.seleniumtesting
    $ seleniumtesting -h



Usage
-----

By itself this product doesn't do anything. You must provide test suites to be run in the form of pkg_resource plugins.
The ``edw.seleniumtesting.sample`` suite is registered in such a way to provide an example.

In your own package, after writing the suite (refer to ``edw.seleniumtesting.sample``) you must the following to the
``setup.py`` of your package (the sample suite is registered in the same way): ::

    entry_points={
        'edw.seleniumtesting': [
            'my.package.test1 = my.package.test1:suite',
            'my.package.test2 = my.package.test2:suite',
            'my.package.test3 = my.package.test3:suite',
            [...]
        ]
    }

and ``pip install my.package``. Modifying the list of tests defined in the ``entry_points`` will require a
re-installation of your package (re-run pip/setuptools/buildout).


To run the ``my.package.test1``, ``my.package.test2`` and ``my.package.test3`` tests in Firefox,
specifying the path to ``geckodriver`` at the default ``1024x768`` resolution: ::

    $ seleniumtesting -v -B firefox -P /usr/bin/geckodriver https://localhost my.package.test1 my.package.test2 my.package.test3


To run all tests in phantomjs in glorious 4K resolution: ::

    $ seleniumtesting -v -B phantomjs -P /usr/bin/phantomjs -sw 3840 -sh 2160 https://localhost

Failed tests and tests that encounter an error will save a screenshot in the current working directory.


Browser arguments
-----------------

Supported for Chrome and Firefox. Can be provided using ``-A``. For example, to run Chrome with the ``--headless`` option: ::

    $ seleniumtesting -v -B chrome -P /usr/lib/chromium-browser/chromedriver -A='--headless' https://localhost


Extra arguments
---------------

Some test suites might make use of extra arguments (e.g. to provide user account credentials).
These can be given using the ``-ea`` or ``--extra-arg`` parameter. For example: ::

  seleniumtesting http://localhost ns.some.test -ea labels login "Login Button" -ea users testuser testuserpwd

The arguments will be passed to ``suite`` as extra_args. Make sure to instantiate your ``BrowserTestCase`` subclass with ``extra_args`` as well. You will then be able to read the args in the test from ``self.extra_args``.

For the example provided above, ``self.extra_args`` will look like this: ::

  {
      'labels': { 'login':    'Login Button' },
      'users':  { 'testuser': 'testuserpwd'  }
  }


Contribute
----------

- Issue Tracker: https://github.com/eaudeweb/edw.seleniumtesting/issues
- Source Code: https://github.com/eaudeweb/edw.seleniumtesting
- Documentation: https://github.com/eaudeweb/edw.seleniumtesting/wiki


License
-------

The project is licensed under the GPLv3.


CHANGES.rst
Changelog
=========


1.1 (2018-04-25)
----------------

- Support for browser options (e.g. --headless)
  [david-batranu]


1.0 (2017-02-17)
----------------

- Initial release.
  [david-batranu]


src/edw/seleniumtesting/main.py
import unittest

from edw.seleniumtesting import common
from edw.seleniumtesting import util


def run(base_url,
        suite,
        resolution=(1024, 768),
        verbosity=1,
        browser='chrome',
        browser_path=None,
        extra_args={}):

    browser = util.get_browser(browser, browser_path)
    browser.set_window_size(*resolution)

    test_runner = unittest.TextTestRunner(
        verbosity=verbosity,
        resultclass=common.BrowserTestResult
    )

    test_suite = suite(browser, base_url, extra_args)
    test_runner.run(test_suite)

    browser.quit()


def run_cli():
    parser = util.build_cli_arguments()
    args = parser.parse_args()

    for test_name in args.test:
        util.validate_test_name(test_name)
        suite = util.ARG_TESTS[test_name]

        run(base_url=args.url,
            suite=suite,
            resolution=(args.screenwidth, args.screenheight),
            verbosity=args.verbose,
            browser=args.browser,
            browser_path=args.browserpath,
            extra_args=dict(args.extra_arg))

if __name__ == '__main__':
    run_cli()


src/edw/seleniumtesting/sample.py
import unittest

from edw.seleniumtesting.common import BrowserTestCase


def suite(browser, base_url, extra_args):
    """ Run with https://google.com/ncr and -ea text "Google Search"
    """
    # setup a new suite
    suite = unittest.TestSuite()

    # Iterate over defined tests and add them to the suite.
    # This is useful if you want to run the same tests on
    # multiple urls starting from the initial page.
    # For example searching Google and running the TestCase tests
    # on each of the results:
    #
    # for result in results:
    #     for name in GoogleTestCase.my_tests():
    #        suite.addTest(GoogleTestCase(name, browser, result.url)

    for name in GoogleTestCase.my_tests():
        testcase = GoogleTestCase(name, browser, base_url, extra_args)
        suite.addTest(testcase)

    return suite


class GoogleTestCase(BrowserTestCase):

    def setUp(self):
        self.browser.get(self.url)

    def test_search_button_exists(self):
        """ 'Google Search' button exists.
        """
        selector = '//*[@value="{}"]'.format(self.extra_args['text'])
        elements = self.browser.find_elements_by_xpath(selector)
        self.assertGreater(len(elements), 0)


src/edw/seleniumtesting/util.py
import argparse
from pkg_resources import iter_entry_points

from selenium import webdriver


DRIVERS = {
    'chrome': webdriver.Chrome,
    'firefox': webdriver.Firefox,
    'phantomjs': webdriver.PhantomJS,
    'edge': webdriver.Edge,
    'ie': webdriver.Ie,
    'safari': webdriver.Safari,
}


MSG_UNKNOWN_TEST = (
    'Unknown test "{}"! '
    'Known tests: "{known_tests}"'
)


ARG_TESTS = {}

for entry_point in iter_entry_points(group='edw.seleniumtesting', name=None):
    ARG_TESTS[entry_point.name] = entry_point.load()


def get_browser(name, path=None):
    browser = DRIVERS[name]
    return browser(executable_path=path) if path else browser()


def validate_test_name(test_name):
    assert test_name in ARG_TESTS, MSG_UNKNOWN_TEST.format(
        test_name, known_tests=', '.join(ARG_TESTS.keys()))


def build_cli_arguments() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description=(
            'Run tests on websites.\n'
            'The given browser webdriver must be in your $PATH\n'
            'or given via the --browserpath option.\n\n'
            'E.g.: chrome: chromedriver, firefox: geckodriver, '
            'edge: MicrosoftWebDriver.exe.'
        )
    )
    parser.add_argument(
        'url', type=str,
        help='Site url, eg: https://digital-agenda-data.eu.'
    )
    parser.add_argument(
        'test', nargs='*', type=str, default=ARG_TESTS.keys(),
        help='Test names, one or more of: "{}". Default: all'.format(
            ', '.join(ARG_TESTS.keys())
        )
    )
    parser.add_argument('-v', '--verbose', action='count', default=1)
    parser.add_argument(
        '-B', '--browser', default='chrome',
        help='Browser to use, known: "{}". Default: chrome'.format(
            ', '.join(DRIVERS.keys())
        )
    )
    parser.add_argument(
        '-P', '--browserpath', default=None,
        help='Custom path to browser executable.'
    )

    parser.add_argument(
        '-sw', '--screenwidth', default=1024,
        help='Screen width. Default: 1024.'
    )

    parser.add_argument(
        '-sh', '--screenheight', default=768,
        help='Screen height. Default: 768.'
    )

    parser.add_argument(
        '-ea', '--extra-arg', nargs=2, action='append',
        default=[],
        help=(
            'Extra arguments passed to suite.\n'
            'You can use this to pass in user credentials or \n'
            'any other dynamic data that can\'t live with the test code.'
        )
    )

    return parser
